src/terraform-provider-paperspace/client.go

```diff
@@ -50,93 +55,173 @@ func SetResData(d *schema.ResourceData, m map[string]interface{}, n string) {
-		ti := resp.Request.TraceInfo()
-		log.Println("[INFO] DNSLookup    :", ti.DNSLookup)
-		log.Println("[INFO] ConnTime     :", ti.ConnTime)
-		log.Println("[INFO] TCPConnTime  :", ti.TCPConnTime)
-		log.Println("[INFO] TLSHandshake :", ti.TLSHandshake)
-		log.Println("[INFO] ServerTime   :", ti.ServerTime)
-		log.Println("[INFO] ResponseTime :", ti.ResponseTime)
-		log.Println("[INFO] TotalTime    :", ti.TotalTime)
-		log.Println("[INFO] IsConnReused :", ti.IsConnReused)
-		log.Println("[INFO] IsConnWasIdle:", ti.IsConnWasIdle)
-		log.Println("[INFO] ConnIdleTime :", ti.ConnIdleTime)
-
-		return nil
-	})
-
-	restyClient.
-		SetHostURL(c.ApiHost).
-		SetHeader("x-api-key", c.ApiKey).
-		SetHeader("Accept", "application/json").
-		SetHeader("Content-Type", "application/json").
-		SetHeader("User-Agent", "terraform-provider-paperspace").
-		SetHeader("ps_client_name", "terraform-provider-paperspace")
+	APIKey     string
+	APIHost    string
+	Region     string
+	HttpClient *http.Client
+}
+
+func (c *ClientConfig) Client() (PaperspaceClient, error) {
+	timeout := 10 * time.Second
+	hc := &http.Client{
+		Timeout: timeout,
+	}
+
+	rt := WithHeader(hc.Transport)
+	rt.Set("x-api-key", c.APIKey)
+	rt.Set("Accept", "application/json")
+	rt.Set("Content-Type", "application/json")
+	rt.Set("User-Agent", "terraform-provider-paperspace")
+	rt.Set("ps_client_name", "terraform-provider-paperspace")
+	hc.Transport = rt
 
 	client := PaperspaceClient{
-		ApiKey:      c.ApiKey,
-		ApiHost:     c.ApiHost,
-		Region:      c.Region,
-		RestyClient: restyClient,
+		APIKey:     c.APIKey,
+		APIHost:    c.APIHost,
+		Region:     c.Region,
+		HttpClient: hc,
 	}
 
 	return client, nil
 }
 
-func LogResponse(reqDesc string, resp *resty.Response, err error) {
-	log.Printf("[INFO] Request: %v", reqDesc)
-	log.Printf("[INFO] Error: %v", err)
-	log.Printf("[INFO] Response Status Code: %v", resp.StatusCode())
-	log.Printf("[INFO] Response Status: %v", resp.Status())
-	log.Printf("[INFO] Response Time: %v", resp.Time())
-	log.Printf("[INFO] Response Received At: %v", resp.ReceivedAt())
-	log.Printf("[INFO] Response Body: %v", resp) // or resp.String() or string(resp.Body())
+// from https://stackoverflow.com/questions/51325704/adding-a-default-http-header-in-go
+type withHeader struct {
+	http.Header
+	rt http.RoundTripper
+}
+
+// WithHeader effectively allows http.Client to have global headers
+func WithHeader(rt http.RoundTripper) withHeader {
+	if rt == nil {
+		rt = http.DefaultTransport
+	}
+
+	return withHeader{
+		Header: make(http.Header),
+		rt:     rt,
+	}
+}
+
+func (h withHeader) RoundTrip(req *http.Request) (*http.Response, error) {
+	for k, v := range h.Header {
+		req.Header[k] = v
+	}
+
+	return h.rt.RoundTrip(req)
+}
+
+func (psc *PaperspaceClient) GetMachine(id string) (body map[string]interface{}, err error) {
+	url := fmt.Sprintf("%s/machines/getMachinePublic?machineId=%s", psc.APIHost, id)
+	req, err := http.NewRequest("GET", url, nil)
```

Comment:
As mentioned in your description New Request, http do, decode calls, logobject, and status code checking (could check for non 200, 201, 204) could be put in a "request" method to DRY up a lot of code.

Reply:
yeah, sounds good for future reference

Reply:
i've gone ahead and created a helper for creating requests and consumed it in for the resourceMachine path, but not yet the others. could y'all give it a look first before i use it elsewhere?